### PR TITLE
[WIP] feat: base config v2 functionality

### PR DESCRIPTION
**Pull request by patrickhulce**
Implements the base functionality for the move to config v2 upon which all other features for the new format can be implemented.

[design doc](https://docs.google.com/document/d/1cN_m591PAQF4tkEf5nJ5M93pI2tM_U64XolmwpOuiUU/edit?usp=sharing)
related #1874 #1875 #1876 #1463 #1826

Missing because WIP (will be fixed before merge):
- ~category description text~
- ~jsdocs~
- ~tests~
- consensus :)

Missing because MVP:
- settings object
- audit/gatherer options
- customizable scoring functions (both proposed are already implemented just hardcoded atm)
- handling of auditResults/artifacts/trace files, planned to be split from config entirely in v2

Current look:

image

### CODE REVIEW

File: lighthouse-core/config/v2/config.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+class ConfigV2 {
+  constructor(configJson, configPath) {
+    if (!configJson) {
+      configJson = defaultConfigJson;
+      configPath = path.resolve(__dirname, defaultConfigPath);
+    }
+
+    // Perform a shallow clone so we can adjust gatherers and audits
+    configJson = Object.assign({}, configJson);
```
Comment (paulirish) [inline lighthouse-core/config/v2/config.js]: I think you'll need to do parse/stringify otherwise the original configJson object can be mutated, right?

Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: I do shallow copies on any children I mutate and whenever this is exposed to outside for potential mutation it's parsed/stringified in `asJson`, this was mainly to preserve the ability to call from javascript with `audits: {custom: {implementation: MyAuditClass}}`

Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: or at least I try to do shallow copies everywhere, we'll see when I add tests :)

File: lighthouse-core/runner.js [hunk 2]
```diff
@@ -134,14 +136,21 @@ class Runner {
...
         return {
           lighthouseVersion: require('../package').version,
           generatedTime: (new Date()).toJSON(),
           initialUrl: opts.initialUrl,
           url: opts.url,
-          audits: formattedAudits,
+          audits: resultsById,
           artifacts: runResults.artifacts,
           runtimeConfig: Runner.getRuntimeConfig(opts.flags),
+          categories,
```
Comment (paulirish) [inline lighthouse-core/runner.js]: should we put this inside a `report` object on the result payload?  or title it `reportCategories` to communicate clearly that it's just presentational?

Comment (patrickhulce) [inline lighthouse-core/runner.js]: Yeah I'm in favor of restructuring this object for v2, should probably add a section to the doc to discuss format

I was thinking something like the following?
```
{
  report: {
    meta: {version, time, url}, // stuff thats currently top-level
    config, // full config used (after extension, cli overrides, etc)
    score, // final overall score
    categories, // heirarchy of computed scores/results
    audits // individual results by audit id
  } // the stuff that gets written to report.json and whose contents can be used to render the HTML report without any additional information
  artifacts: {gatherers, trace, perfLog/devtoolsMessages} // used to quickly re-run analysis/perform later deep dives/debug/etc
}
```

File: lighthouse-core/config/v2/config.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  static objectToList(object) {
+    return Object.keys(object).reduce((list, id) => {
+      list.push(Object.assign({id}, object[id]));
+      return list;
+    }, []).sort((itemA, itemB) => itemA.order - itemB.order);
+  }
+
+  static getGathererImplementations(gatherersObject) {
```
Comment (paulirish) [inline lighthouse-core/config/v2/config.js]: this and `getAuditImplementations` appear to be the same. You leaving them WET for now expecting them to change in the future?

File: lighthouse-core/config/v2/config.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        usedGathererNames.add(gatherer.implementation.name);
+        return gatherer;
+      });
+
+      return Object.assign({}, definition, {gatherers: foundGatherers});
+    });
+
+    if (gathererIds.size !== usedGathererIds.size) {
+      const unused = _subtract(gathererIds, usedGathererIds);
+      log.warn('config', `Gatherers are unused: ${unused.join(', ')}`);
```
Comment (paulirish) [inline lighthouse-core/config/v2/config.js]: is this a bad thing? or just noting they are being culled from the working config based on the selection.

I'm fine with `warn` but the text could be more neutral to indicate its okay. (I'm expecting that this will log a lot when you do `--perf`)

**Comment by paulirish** [comment]
Super excited about this work.  :)

On the whole I'm super excited about the new declarative config. The v2/config is a big upgrade over the current config.

Let's get some tests in place. IMO Category text can wait for another PR. 

I spoke with @brendankenny  and we're thinking of him enabling the v2 flag which enables the DOM based report renderer. This will start as an empty page and we'll build it out from there.

File: lighthouse-core/config/v2/config.js (hunk 4 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: Seemed like a good thing to warn about in case you misname or something, this function is likely to experience the most change when we add `only`/`skip` etc, so I'd say we can re-evaluate or just remove in the filtered case.

File: lighthouse-core/config/v2/config.js (hunk 3 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: So they are, haha I'll DRY up. It was copypasta I expected to modify but didn't need to in the end

File: lighthouse-core/config/v2/config.js (hunk 1 above)
Comment (brendankenny) [inline lighthouse-core/config/v2/config.js]: > or at least I try to do shallow copies everywhere, we'll see when I add tests :)

test will be good :) `assert.notDeepStrictEqual` might be sufficient, but we could also do something with deep `Object.freeze` and see what throws (in strict mode)

Especially important for the browserified/lighthouse-as-a-module case since when loading the config json via `require()` any changes to the original object is persisted to the next use of it.

File: lighthouse-core/report/v2/report-generator.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      return Object.assign({}, category, {children, score});
+    });
+
+    const scoreByCategory = categories.reduce((scores, category) => {
+      scores[category.id] = category.score;
+      return scores;
+    }, {});
+
+    const score = this._geometricMean([
+      {score: scoreByCategory.pwa, weight: 5},
```
Comment (brendankenny) [inline lighthouse-core/report/v2/report-generator.js]: hard coding category names and weights here seems like a mistake. Were you waiting to add a `scoring` property? Waiting is good, but can you just add `weight` to the category objects and skip this? (I'll comment in the design doc)

I also think using `_geometricMean` here is premature until we've had a real discussion on scoring, but happy to have it here for now :P

File: lighthouse-core/runner.js (hunk 2 above)
Comment (brendankenny) [inline lighthouse-core/runner.js]: agree that top level `categories` here is a bit confusing. Can we call it `reportCategories` until we settle on a result object shape?

File: lighthouse-core/config/v2/default.json [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+      "path": "dobetterweb/tags-blocking-first-paint"
+    },
+    "websql": {
+      "path": "dobetterweb/websql"
+    }
+  },
+
+  "passes": {
+    "defaultPass": {
+      "order": 10,
```
Comment (brendankenny) [inline lighthouse-core/config/v2/default.json]: can we kick `order` from the MVP so we can discuss it in a separate PR? I'm still not really a fan and with no motivating code that needs it yet there's no rush

File: lighthouse-core/config/v2/config.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    // Resolve the paths of audits and gatherers
+    configJson.gatherers = ConfigV2.resolvePaths(configJson.gatherers, configPath, [
+      path.join(__dirname, '../../gather/gatherers'),
+    ]);
+    configJson.audits = ConfigV2.resolvePaths(configJson.audits, configPath, [
+      path.join(__dirname, '../../audits'),
+    ]);
+
+    // Extend only after our paths have been resolved
+    configJson = ConfigV2.extendIfNecessary(configJson, configPath);
```
Comment (brendankenny) [inline lighthouse-core/config/v2/config.js]: I kind of want to punt extensions from the MVP as well, that's going to take some thought to get it right, but it looks like you're just matching current behavior here?

**Comment by brendankenny** [comment]
I'd really like to clear up the audit/gatherer id/name/classname/key-in-various-objects distinctions a bit more.

- What is the relationship between the keys in the `gatherers` top level object and the `gatherers` object in each pass? Do the pass gatherer keys reference the keys in the top level object?
- Same thing for the `audits` top-level object keys vs the `id` values in each category's children array (I believe I asked this in a comment)
- Is there a way to use a gatherer twice with different input? This is hard since we use `artifacts[classname]` currently to identify the output of each gatherer, so we may want to table this for now.
- Is there a way to use an audit twice? Since we're letting audits take input with this new version this is going to be a great thing. Ideally I could do something like
```js
"time-to-interactive-cold": {
  "path": "time-to-interactive",
  "inputTrace": "firstPass"
},
"time-to-interactive-warm": {
  "path": "time-to-interactive",
  "inputTrace": "laterPass"
}
```
(modulo better naming/clearer handling of path or whatever, etc)
and then being able to refer to `time-to-interactive-cold` and `time-to-interactive-warm` as different things in my categories. This would require using the keys as ids though, and not something on the audit itself.

I haven't fully thought that through yet, so there may be issues with that :)

So, basically, it would be sweet to simplify the id/key/path thing a bit more and figure out for sure how we're going to allow single audits multiple times (bonus points for gatherers multiple times, though that seems a lot messier).

**Comment by brendankenny** [comment]
kind of shitty idea, everyone feel free to disagree: to iterate on some of this a bit more but not block the new report, we could make even more of an MVP by making the config v2 exactly the same as v1 except have `report.categories` instead of `aggregations`. I think everyone is already in agreement on the `report.categories` stuff here.

`config.js` probably wouldn't need to change much from master and the report generation/scoring could stay as it is in this PR. It's mostly v2/config.js that wouldn't be needed yet.

Just to be clear, this is all really good and would just get rolled over to the next change, I'm just thinking about this because I don't want to slow us way down by making us talk about all these things.

...? :)

File: lighthouse-core/config/v2/config.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
+const path = require('path');
+const log = require('../../lib/log');
+
+const defaultConfigPath = './default.json';
+const defaultConfigJson = require(defaultConfigPath);
+
+const _flatten = arr => [].concat(...arr);
+const _subtract = (setA, setB) => Array.from(setA).filter(x => !setB.has(x));
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: sure, and it did at first but it never gets used as a set, just as an array, I'll rename to `_differenceAsArray`

File: lighthouse-core/config/v2/config.js (hunk 7 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: > it looks like you're just matching current behavior here?

The intent behind several of the config changes was specifically to enable clear and simple extension not needing logic or fancy behavior of any sort, so this is the final behavior as far as I'm concerned. That's what the structure of the new config enables :)

File: lighthouse-core/config/v2/default.json [hunk 9]
```diff
@@ -0,0 +1,308 @@
+{
+  "version": 2,
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: Agreed we should have verification of the config format but we explicitly are attempting to support two config versions at once to start as we iterate, so I think the version has value as a clue of which path to head toward. A try until it works approach with validation and error messages greatly complicates the MVP. How about `configVersion` until it becomes the default and we nuke the field entirely and rely on validation only?

File: lighthouse-core/config/v2/default.json [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    "content-width": {},
+    "service-worker": {},
+    "offline": {},
+    "http-redirect": {},
+    "html-without-javascript": {},
+    "chrome-console-messages": {},
+    "styles": {},
+    "css-usage": {},
+    "all-event-listeners": {
+      "path": "dobetterweb/all-event-listeners"
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: Not sure what you're trying to say here. IMO, this is actually a step in the direction of decoupling our arbitrary repo folder structure from magic happening in the config. What leads you to believe our repo organization process is now inherent to the gatherer now?

File: lighthouse-core/config/v2/default.json (hunk 6 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: sure

File: lighthouse-core/config/v2/default.json [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    "manifest-icons-min-192": {},
+    "manifest-icons-min-144": {},
+    "manifest-name": {},
+    "manifest-short-name": {},
+    "manifest-short-name-length": {},
+    "manifest-start-url": {},
+    "theme-color-meta": {},
+    "content-width": {},
+    "deprecations": {},
+    "aria-allowed-attr": {
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: For audits, as with each of these toplevel objects, the keys are the `id` (currently `audit.meta.name`) of the audit (rather than have it redundantly specified inside the object as well). There's really no need for audits to have category or name in their implementation now either.

File: lighthouse-core/config/v2/default.json [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+      "path": "dobetterweb/uses-passive-event-listeners"
+    }
+  },
+
+  "report": {
+    "categories": {
+      "pwa": {
+        "name": "Progressive Web App",
+        "description": "",
+        "children": [
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: `children` was used forseeing the whole `subcategory` discussion occurring and the likely scenario that we will need to add categories as children of other categories. Keeping it generic with children, IMO, leaves open the possibility without a breaking change. In the proposal, each child has a `type: "audit"|"category"|"some-newfangled-thing"` but since all we use are audits today it seemed unnecessary.

File: lighthouse-core/gather/gather-runner.js [hunk 13]
```diff
@@ -431,6 +431,11 @@ class GatherRunner {
   static instantiateGatherers(passes, rootPath) {
     return passes.map(pass => {
       pass.gatherers = pass.gatherers.map(gatherer => {
+        if (gatherer.implementation) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js]: sure

File: lighthouse-core/report/v2/report-generator.js (hunk 5 above)
Comment (patrickhulce) [inline lighthouse-core/report/v2/report-generator.js]: Yeah I was waiting to introduce scoring methods and since nothing scoring wise has changed despite lots of talk the past few months I thought I'd poke the bear and see what came of it :P

File: lighthouse-core/runner.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/runner.js]: done

**Comment by patrickhulce** [comment]
> audit/gatherer id/name/classname/key-in-various-objects

I seem to have failed in config clarity given that I think the biggest win of config v2 is getting rid all of most of this nonsense and moving everything to IDs (apart from artifact name being class name...still need to hash that one out). The keys for *all toplevel objects* (gatherers, audits, passes) are IDs of that object. This ID should how everything else in the config refers to that particular object. Examples below.

> What is the relationship between the keys in the gatherers top level object and the gatherers object in each pass? Do the pass gatherer keys reference the keys in the top level object?

The strings specified in a passes `gatherers` property are the IDs of gatherers. The long-term view was that these wouldn't need to be strings but potentially objects specifying a gatherer ID and potential options (for example collecting similar data across passes, or with slightly different settings).

> Same thing for the audits top-level object keys vs the id values in each category's children array (I believe I asked this in a comment)

Responded in comment too, but same story. These strings are the IDs of the audits.

> Is there a way to use a gatherer twice with different input? This is hard since we use artifacts[classname] currently to identify the output of each gatherer, so we may want to table this for now.

Correcto, as mentioned above still need to hash this out, but my big controversial proposal will be to start storing artifacts by gatherer ID, ignore class name completely, and move requiredArtifacts to an input property of an audit (with a defaults suggested by impl so config doesn't balloon with unnecessary wiring).....buuuuuuut one major change at a time :)

> Is there a way to use an audit twice? Since we're letting audits take input with this new version this is going to be a great thing. Ideally I could do something like

Bingo :)

> This would require using the keys as ids though, and not something on the audit itself.

yes! this is exactly what I'm trying to move to.

**Comment by patrickhulce** [comment]
> making the config v2 exactly the same as v1 except have report.categories instead of aggregations

😢 😢 

Seriously though, IMO there's no need to have these set in stone until we actually flip the switch and kill config v1 making v2 the default. If we have buyer's remorse about passes as object/whatever, I don't see why we couldn't change it all around tomorrow (though I obviously hope major points could be ironed out in doc stage 😄 )

File: lighthouse-core/config/v2/default.json [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+        "domstats",
+        "optimized-images",
+        "tags-blocking-first-paint",
+        "websql"
+      ]
+    }
+  },
+
+  "audits": {
+    "is-on-https": {},
```
Comment (ebidel) [inline lighthouse-core/config/v2/default.json]: shall we alphabetize all of these listings?

File: lighthouse-core/config/v2/default.json (hunk 12 above)
Comment (ebidel) [inline lighthouse-core/config/v2/default.json]: I'm in favor of keeping "audits" for now. I read this block as "this category contains/runs these audits".  That naming is consistent with the list of "audits" above and is probably easier for contributors to make the linkage when adding new audits.

File: lighthouse-core/config/v2/default.json [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    },
+    "uses-http2": {
+      "path": "dobetterweb/uses-http2"
+    },
+    "uses-passive-event-listeners": {
+      "path": "dobetterweb/uses-passive-event-listeners"
+    }
+  },
+
+  "report": {
```
Comment (ebidel) [inline lighthouse-core/config/v2/default.json]: I still thinks it's worth exploring modularity but am alright with adding a `report` section. 

IOW,`report` could be it's own file (`config/report.json`). In that way, we keep configuration separate from output generation. As as an example, I could see users wanting to use our settings.json with their own report.json to create a custom report. They wouldn't have to change a mega file. That said, maybe your extension model smooths over that nasty....?

File: lighthouse-core/config/v2/default.json [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+          {"id": "manifest-icons-min-192", "weight": 1},
+          {"id": "manifest-exists", "weight": 1},
+          {"id": "theme-color-meta", "weight": 1},
+          {"id": "viewport", "weight": 1},
+          {"id": "content-width", "weight": 1}
+        ]
+      },
+      "performance": {
+        "name": "Performance",
+        "description": "",
```
Comment (ebidel) [inline lighthouse-core/config/v2/default.json]: Might as well port over the descriptions from v1 while we're here.

File: lighthouse-core/report/v2/report-generator.js (hunk 5 above)
Comment (ebidel) [inline lighthouse-core/report/v2/report-generator.js]: Anything related to scoring does feel like a future PR. There's enough to bit off in here with all the other changes :)

File: lighthouse-core/config/v2/config.js [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        path.resolve(configPath, rawPath), // for relative config usage
+        path.resolve(process.cwd(), rawPath), // for node module usage
+      ]);
+      const resolvedPath = ConfigV2._tryResolveUntilSuccess(possiblePaths);
+      object[key] = Object.assign({}, object[key], {path: resolvedPath});
+    });
+    return object;
+  }
+
+  static objectToList(object) {
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: mainly for testing since it did do fancier things like ordering, but that's been nixed for a future PR and didn't seem worth moving

File: lighthouse-core/config/v2/config.js [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    const extendedConfig = new ConfigV2(extendedConfigJson, extendedConfigPath);
+    return ConfigV2._mergeObjects(extendedConfig.asJson(), configJson);
+  }
+
+  static resolvePaths(object, configPath, searchPaths = []) {
+    object = Object.assign({}, object);
+    Object.keys(object).forEach(key => {
+      // We don't need to resolve objects that already have
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/config.js]: fixed

File: lighthouse-core/config/v2/default.json [hunk 19]
```diff
@@ -0,0 +1,304 @@
+{
+  "configVersion": 2,
+
+  "gatherers": {
```
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: yep :) lists all gatherers with their ids and paths (if different from id) plus allowing options and multiple version of the same implementation later on down the road

File: lighthouse-core/config/v2/default.json (hunk 15 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: I'm definitely not opposed to separating the presentation config from the runtime config at some point, easy extension eases the immediate pain IMO.

File: lighthouse-core/config/v2/default.json (hunk 16 above)
Comment (patrickhulce) [inline lighthouse-core/config/v2/default.json]: done
